File: src/algebra2_rational_expressions_hard.py
"""
algebra2 - rational_expressions (hard)
Generated: 2026-02-22T04:24:03.925909
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Simplifying complex rational expressions with multiple terms
        a1 = nonzero(-5, 5)
        b1 = nonzero(-5, 5)
        a2 = nonzero(-5, 5)
        b2 = nonzero(-5, 5)
        
        # Create (a1*x + b1)/(x^2 - c^2) + (a2*x + b2)/(x^2 - c^2)
        c = randint(2, 5)
        denom = x**2 - c**2
        
        numer1 = a1*x + b1
        numer2 = a2*x + b2
        
        combined_numer = expand(numer1 + numer2)
        ans = simplify(combined_numer / denom)
        
        return problem(
            question=f"Simplify: $\\frac{{{latex(numer1)}}}{{{latex(denom)}}} + \\frac{{{latex(numer2)}}}{{{latex(denom)}}}$",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"The denominators are the same: ${latex(denom)}$",
                f"Add the numerators: ${latex(numer1)} + ({latex(numer2)}) = {latex(combined_numer)}$",
                f"Combined: $\\frac{{{latex(combined_numer)}}}{{{latex(denom)}}}$",
                f"Factor denominator: ${latex(factor(denom))}$",
                f"Simplified: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 2:
        # Complex fraction (fraction divided by fraction)
        r1 = nonzero(-4, 4)
        r2 = nonzero(-4, 4)
        while r1 == r2:
            r2 = nonzero(-4, 4)
        
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        
        # (a*x + b)/(x - r1) divided by (x - r2)/(x + r1)
        numer_top = a*x + b
        denom_top = x - r1
        numer_bottom = x - r2
        denom_bottom = x + r1
        
        # Answer is (a*x + b)/(x - r1) * (x + r1)/(x - r2)
        result = simplify((numer_top * denom_bottom) / (denom_top * numer_bottom))
        ans = result
        
        return problem(
            question=f"Simplify: $\\frac{{\\frac{{{latex(numer_top)}}}{{{latex(denom_top)}}}}}{{\\frac{{{latex(numer_bottom)}}}{{{latex(denom_bottom)}}}}}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Rewrite as multiplication by the reciprocal:",
                f"$\\frac{{{latex(numer_top)}}}{{{latex(denom_top)}}} \\cdot \\frac{{{latex(denom_bottom)}}}{{{latex(numer_bottom)}}}$",
                f"Multiply: $\\frac{{({latex(numer_top)})({latex(denom_bottom)})}}{{({latex(denom_top)})({latex(numer_bottom)})}}$",
                f"Simplified: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 3:
        # Subtracting rational expressions with different denominators
        r1 = nonzero(-5, 5)
        r2 = nonzero(-5, 5)
        while r1 == r2 or r1 == -r2:
            r2 = nonzero(-5, 5)
        
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        
        # a/(x - r1) - b/(x - r2)
        numer1 = a
        denom1 = x - r1
        numer2 = b
        denom2 = x - r2
        
        # LCD is (x - r1)(x - r2)
        lcd = expand((x - r1) * (x - r2))
        new_numer1 = expand(a * (x - r2))
        new_numer2 = expand(b * (x - r1))
        combined = expand(new_numer1 - new_numer2)
        
        ans = simplify(combined / lcd)
        
        return problem(
            question=f"Simplify: $\\frac{{{a}}}{{{latex(denom1)}}} - \\frac{{{b}}}{{{latex(denom2)}}}$",
            answer=ans,
            difficulty=(1700, 1800),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Find LCD: $({latex(denom1)})({latex(denom2)}) = {latex(lcd)}$",
                f"Rewrite first fraction: $\\frac{{{a} \\cdot ({latex(x - r2)})}}{{{latex(lcd)}}} = \\frac{{{latex(new_numer1)}}}{{{latex(lcd)}}}$",
                f"Rewrite second fraction: $\\frac{{{b} \\cdot ({latex(x - r1)})}}{{{latex(lcd)}}} = \\frac{{{latex(new_numer2)}}}{{{latex(lcd)}}}$",
                f"Subtract: $\\frac{{{latex(new_numer1)} - ({latex(new_numer2)})}}{{{latex(lcd)}}} = \\frac{{{latex(combined)}}}{{{latex(lcd)}}}$",
                f"Simplified: ${latex(ans)}$"
            )
        )
    
    elif problem_type == 4:
        # Solving rational equations
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = nonzero(-6, 6)
        r = nonzero(-5, 5)
        
        # (a*x + b)/(x - r) = c
        # Solving: a*x + b = c*(x - r)
        # a*x + b = c*x - c*r
        # a*x - c*x = -c*r - b
        # x(a - c) = -c*r - b
        
        if a == c:
            a += 1
        
        sol = (-c*r - b) / (a - c)
        ans = simplify(sol)
        
        return problem(
            question=f"Solve for $x$: $\\frac{{{latex(a*x + b)}}}{{{latex(x - r)}}} = {c}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Multiply both sides by $({latex(x - r)})$:",
                f"${latex(a*x + b)} = {c}({latex(x - r)})$",
                f"${latex(a*x + b)} = {latex(expand(c*(x - r)))}$",
                f"${latex(a*x)} - {latex(c*x)} = {-c*r} - {b}$",
                f"${latex((a-c)*x)} = {-c*r - b}$",
                f"$x = {latex(ans)}$"
            )
        )
    
    else:
        # Adding three rational expressions with different denominators
        r1 = randint(1, 4)
        r2 = randint(1, 4)
        while r2 == r1:
            r2 = randint(1, 4)
        r3 = randint(1, 4)
        while r3 == r1 or r3 == r2:
            r3 = randint(1, 4)
        
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(-3, 3)
        
        # a/(x - r1) + b/(x - r2) + c/(x - r3)
        denom1 = x - r1
        denom2 = x - r2
        denom3 = x - r3
        
        lcd = expand((x - r1) * (x - r2) * (x - r3))
        
        term1 = expand(a * (x - r2) * (x - r3))
        term2 = expand(b * (x - r1) * (x - r3))
        term3 = expand(c * (x - r1) * (x - r2))
        
        combined = expand(term1 + term2 + term3)
        ans = simplify(combined / lcd)
        
        return problem(
            question=f"Simplify: $\\frac{{{a}}}{{{latex(denom1)}}} + \\frac{{{b}}}{{{latex(denom2)}}} + \\frac{{{c}}}{{{latex(denom3)}}}$",
            answer=ans,
            difficulty=(1750, 1850),
            topic="algebra2/rational_expressions",
            solution=steps(
                f"Find LCD: ${latex(lcd)}$",
                f"First term: $\\frac{{{a}({latex((x-r2)*(x-r3))})}}{{{latex(lcd)}}} = \\frac{{{latex(term1)}}}{{{latex(lcd)}}}$",
                f"Second term: $\\frac{{{b}({latex((x-r1)*(x-r3))})}}{{{latex(lcd)}}} = \\frac{{{latex(term2)}}}{{{latex(lcd)}}}$",
                f"Third term: $\\frac{{{c}({latex((x-r1)*(x-r2))})}}{{{latex(lcd)}}} = \\frac{{{latex(term3)}}}{{{latex(lcd)}}}$",
                f"Combined numerator: ${latex(combined)}$",
                f"Answer: ${latex(ans)}$"
            )
        )

emit(generate())

File: src/algebra2_rational_expressions_hard.jl
# algebra2 - rational_expressions (hard)
# Generated: 2026-03-08T20:15:45.417115

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("algebra2/rational_expressions")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Simplify rational expression with polynomial factoring
        r1, r2 = nonzero(-8, 8), nonzero(-8, 8)
        r3 = nonzero(-8, 8)
        while r3 == r1 || r3 == r2
            r3 = nonzero(-8, 8)
        end
        
        numerator = expand((x - r1) * (x - r2))
        denominator = expand((x - r1) * (x - r3))
        
        simplified_num = x - r2
        simplified_den = x - r3
        
        ans = simplified_num / simplified_den
        
        problem(
            question="Simplify the rational expression: \$\\frac{$(tex(numerator))}{$(tex(denominator))}\$",
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor numerator: \$$(tex((x - r1) * (x - r2)))\$",
                "Factor denominator: \$$(tex((x - r1) * (x - r3)))\$",
                "Cancel common factor \$(x - $(r1))\$",
                sol("Answer", ans),
                "Restriction: \$x \\neq $(r1), $(r3)\$"
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Add/subtract rational expressions with different denominators
        r1, r2 = nonzero(-10, 10), nonzero(-10, 10)
        while r2 == r1
            r2 = nonzero(-10, 10)
        end
        
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        
        # a/(x-r1) + b/(x-r2)
        num1 = a
        den1 = x - r1
        num2 = b
        den2 = x - r2
        
        # Common denominator: (x-r1)(x-r2)
        common_den = expand((x - r1) * (x - r2))
        combined_num = expand(a * (x - r2) + b * (x - r1))
        
        ans = combined_num / common_den
        
        operation = choice(["+", "-"])
        if operation == "-"
            num2 = -b
            combined_num = expand(a * (x - r2) - b * (x - r1))
            ans = combined_num / common_den
        end
        
        problem(
            question="Simplify: \$\\frac{$(a)}{$(tex(den1))} $(operation) \\frac{$(abs(b))}{$(tex(den2))}\$",
            answer=ans,
            difficulty=(1900, 2100),
            solution=steps(
                "Find common denominator: \$$(tex(common_den))\$",
                "Rewrite first fraction: \$\\frac{$(a) \\cdot $(tex(x - r2))}{$(tex(common_den))}\$",
                "Rewrite second fraction: \$\\frac{$(operation == "+" ? b : -b) \\cdot $(tex(x - r1))}{$(tex(common_den))}\$",
                "Combine numerators: \$$(tex(combined_num))\$",
                sol("Answer", ans)
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Solve rational equation
        r1 = nonzero(-10, 10)
        r2 = nonzero(-10, 10)
        while r2 == r1
            r2 = nonzero(-10, 10)
        end
        
        ans = nonzero(-15, 15)
        while ans == r1 || ans == r2
            ans = nonzero(-15, 15)
        end
        
        a = nonzero(-6, 6)
        # a/(x-r1) = (x-r2)/(x-r1)
        # Cross multiply: a(x-r1) = (x-r2)(x-r1)
        # We want solution to be ans
        # So: a = x - r2 evaluated at x = ans
        a = ans - r2
        
        lhs = a / (x - r1)
        rhs = (x - r2) / (x - r1)
        
        problem(
            question="Solve for x: \$\\frac{$(a)}{$(tex(x - r1))} = \\frac{$(tex(x - r2))}{$(tex(x - r1))}\$",
            answer=ans,
            difficulty=(1850, 2050),
            solution=steps(
                sol("Given", lhs ~ rhs),
                "Since denominators are equal, numerators must be equal",
                sol(a ~ (x - r2)),
                "Solve for x: \$x = $(a + r2)\$",
                "Check: \$x \\neq $(r1)\$ (restriction)",
                sol("Answer", ans)
            ),
            time=200
        )
        
    elseif problem_type == 4
        # Complex fraction simplification
        r1, r2 = nonzero(-7, 7), nonzero(-7, 7)
        while r2 == r1
            r2 = nonzero(-7, 7)
        end
        
        a = nonzero(-4, 4)
        b = nonzero(-4, 4)
        
        # (a + b/x) / (1 - r1/x) = (ax + b) / (x - r1)
        numerator_top = a*x + b
        numerator_bottom = x - r1
        
        ans = numerator_top / numerator_bottom
        
        problem(
            question="Simplify the complex fraction: \$\\frac{$(a) + \\frac{$(b)}{x}}{1 - \\frac{$(r1)}{x}}\$",
            answer=ans,
            difficulty=(2000, 2200),
            solution=steps(
                "Multiply numerator and denominator by \$x\$",
                "Numerator becomes: \$$(a)x + $(b)\$",
                "Denominator becomes: \$x - $(r1)\$",
                sol("Answer", ans),
                "Restriction: \$x \\neq 0, $(r1)\$"
            ),
            time=220
        )
        
    else
        # Rational function domain with multiple restrictions
        r1 = nonzero(-12, 12)
        r2 = nonzero(-12, 12)
        while r2 == r1
            r2 = nonzero(-12, 12)
        end
        r3 = nonzero(-12, 12)
        while r3 == r1 || r3 == r2
            r3 = nonzero(-12, 12)
        end
        
        # Function: (x - r3) / [(x - r1)(x - r2)]
        numerator = x - r3
        denominator = expand((x - r1) * (x - r2))
        
        restrictions = sort([r1, r2])
        
        # Domain is all reals except r1, r2
        ans = "(-\\infty, $(restrictions[1])) \\cup ($(restrictions[1]), $(restrictions[2])) \\cup ($(restrictions[2]), \\infty)"
        
        problem(
            question="Find the domain of \$f(x) = \\frac{$(tex(numerator))}{$(tex(denominator))}\$. Express in interval notation.",
            answer=ans,
            difficulty=(2100, 2300),
            answer_type="interval",
            solution=steps(
                sol("Given", numerator / denominator),
                "Factor denominator: \$$(tex((x - r1) * (x - r2)))\$",
                "Denominator equals zero when \$x = $(r1)\$ or \$x = $(r2)\$",
                "These values must be excluded from the domain",
                sol("Domain", ans)
            ),
            time=180
        )
    end
end